File: src/Web/JavaScriptJS/patterns/microkernel.js
const assert = require("node:assert/strict");
/** Keeps a small core stable while plugins extend named capabilities. */
function run() {
  const plugins = new Map();
  const kernel = {
    register(name, plugin) { plugins.set(name, plugin); },
    execute(name, input) { return plugins.get(name)(input); },
  };
  kernel.register("uppercase", (value) => value.toUpperCase());
  assert.equal(kernel.execute("uppercase", "genkidama"), "GENKIDAMA");
  return "microkernel";
}
module.exports = { run };
if (require.main === module) run();

File: src/DataScience/MATLAB/microkernel.m
function result = microkernel
%MICROKERNEL Keep a small core and extend behavior through registered plugins.
kernel = struct();
kernel.plugins.double = @(value) value * 2;
kernel.plugins.square = @(value) value * value;
result = struct( ...
    "double", invokePlugin(kernel, "double", 4), ...
    "square", invokePlugin(kernel, "square", 4));
end

function value = invokePlugin(kernel, name, input)
if ~isfield(kernel.plugins, char(name))
    error("Unknown plugin: %s", name);
end
plugin = kernel.plugins.(char(name));
value = plugin(input);
end
